src/PolyType/Utilities/AggregatingTypeShapeProvider.cs

```diff
@@ -0,0 +1,63 @@
+﻿using PolyType.Abstractions;
+
+namespace PolyType.Utilities;
+
+/// <summary>
+/// An <see cref="ITypeShapeProvider"/> that returns the first shape found
+/// from iterating over a given list of providers.
+/// </summary>
+/// <remarks>
+/// <para>
+/// This aggregator implements the chain of responsibility pattern.
+/// </para>
+/// <para>
+/// Whether this aggregator caches the results of the providers is undefined,
+/// and should be non-observable assuming the providers themselves do not
+/// change their return values over time.
+/// </para>
+/// </remarks>
+public sealed class AggregatingTypeShapeProvider : ITypeShapeProvider
+{
+    private readonly ITypeShapeProvider[] _providers;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AggregatingTypeShapeProvider"/> class.
+    /// </summary>
+    /// <param name="providers">The sequence of providers to solicit for type shapes.</param>
+    public AggregatingTypeShapeProvider(params ITypeShapeProvider[] providers)
+    {
+        if (providers is null)
+        {
+            throw new ArgumentNullException(nameof(providers));
+        }
+
+        foreach (ITypeShapeProvider provider in providers)
+        {
+            if (provider is null)
+            {
+                throw new ArgumentException("One or more providers are null.", nameof(providers));
+            }
+        }
+
+        _providers = providers.ToArray();
+    }
+
+    /// <summary>
+    /// Gets the list of providers making up this aggregate provider.
+    /// </summary>
+    public IReadOnlyList<ITypeShapeProvider> Providers => _providers;
```

Comment:
This allows mutation of the underlying array if you cast it. I would recommend changing the type of `_providers` to `IReadOnlyList<ITypeShapeProvider>` and then populate that in the constructor using collection literals:
```C#
_providers = [ ..providers];
```

Reply:
Yeah. It could have used `ImmutableArray<T>` I suppose but maybe it can't just assume that it's available.

Reply:
Ya, I need to reference S.C.I. on net472 to get ImmutableArray. Per [this comment](https://github.com/eiriktsarpalis/PolyType/pull/107#discussion_r1929872467).

However, defining the field to be an interface type means that `foreach` is no longer alloc free. Sure, I could rewrite my `foreach` loop to a `for` loop and use the indexer, but having to remember to avoid `foreach` because the field is typed as an interface instead of a concrete type seems like a fairly precarious position. Generally, and I think there are analyzers to help encourage this, private fields ought to be concrete types to make allocation less likely.

Reply:
FWIW I opted for using the compiler-generated IReadOnlyList over ReadOnlyCollection<T> elsewhere in this project. So opting for a dependency on System.Collections.ObjectModel should if anything increase the static footprint of trimmed apps.